### PR TITLE
Add Azure AD capability to module

### PR DESCRIPTION
**Pull request by madmatt**
This is the beginnings of a PR to add Azure AD capability to the module. Still need to determine if it's backwards compatible (I think it is).

It has been tested against a stock standard Azure AD trial account and works well, provided the new instructions are followed by the Azure AD administrator.

### CODE REVIEW

**Comment by madmatt** [comment]
Have just removed the `[WIP]` tag, this is being used in a few different production sites now and seems to work well across environments. I haven't yet been able to review these changes against an ADFS target to make sure nothing has regressed, and I don't expect to have the opportunity to do this anytime soon unfortunately.

Perhaps we can release this as a new major version so it doesn't break existing installs that run `composer update`.

This can now be reviewed.

File: src/Services/SAMLConfiguration.php [hunk 1]
```diff
@@ -52,6 +52,25 @@ class SAMLConfiguration
...
     private static $authn_contexts;
 
+    /**
+     * @config
+     * @var bool Whether or not we expect to receive a binary NameID from the IdP. We expect to receive a binary NameID
+     * from ADFS, but don't expect it from Azure AD or most other SAML implementations that provide GUIDs.
+     *
+     * Defaults to true to preserve backwards compatibility (ADFS).
+     */
+    private static $expect_binary_nameid;
```
Comment (NightJar) [inline src/Services/SAMLConfiguration.php]: Docblock states that this value defaults to true, however there is no default value, thus it is false (`NULL`).
If it is true from yml sources I think it best also reflected here in the "hard truth", coinciding with the docblock.

**Review by NightJar — APPROVED** [review]
Code looks good, and love the amount of documentation. One small nitpick, but I won't let that block a merge.

**Comment by madmatt** [comment]
Thanks @NightJar, I've fixed that up now, and also added some extra functionality. Will fix the linting errors at some point shortly, but could you possible review my changes above? I'll squash before merge, but figured you'd prefer to only look at the new commits for now ;-)

**Comment by madmatt** [comment]
Without spending a lot of time faking SAML responses I don't see a nice / easy way to test the SAMLController->acs() method, and that's outta scope for me right now. Other than that issue with codecov.io, I think this is ready for merge.

@NightJar Any chance of a quick review of the latest commits, and if you're happy I'll squash and you can merge, then I can release? Thanks!

**Review by NightJar — APPROVED** [review]

**Pull request merged**